### ISSUE TITLE
Failed to create project test1

### ISSUE THREAD

**Issue by elharo** [GoogleCloudPlatform/google-cloud-eclipse#1568]
![image](https://cloud.githubusercontent.com/assets/1005544/23795498/fa53cc8e-0563-11e7-8d6b-23bdee099679.png)

At head while trying to create standard project

```
Java Model Exception: Java Model Status [test1 does not exist]
	at org.eclipse.jdt.internal.core.JavaElement.newNotPresentException(JavaElement.java:556)
	at org.eclipse.jdt.internal.core.JavaModelManager.getPerProjectInfoCheckExistence(JavaModelManager.java:2307)
	at org.eclipse.jdt.internal.core.JavaProject.getPerProjectInfo(JavaProject.java:1909)
	at org.eclipse.jdt.internal.core.JavaProject.getRawClasspath(JavaProject.java:1931)
	at com.google.cloud.tools.eclipse.appengine.newproject.CreateAppEngineWtpProject.addJunit4ToClasspath(CreateAppEngineWtpProject.java:121)
	at com.google.cloud.tools.eclipse.appengine.newproject.CreateAppEngineWtpProject.execute(CreateAppEngineWtpProject.java:108)
	at org.eclipse.ui.actions.WorkspaceModifyOperation$1.run(WorkspaceModifyOperation.java:108)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2241)
	at org.eclipse.ui.actions.WorkspaceModifyOperation.run(WorkspaceModifyOperation.java:130)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)
```

**Comment by elharo** [comment]
I have not been immediately able to reproduce this. Perhaps a thread issue?

```
  private static void addJunit4ToClasspath(IProgressMonitor monitor, IProject newProject)
      throws CoreException {
    IJavaProject javaProject = JavaCore.create(newProject);
    IClasspathAttribute nonDependencyAttribute =
        UpdateClasspathAttributeUtil.createNonDependencyAttribute();
    IClasspathEntry junit4Container = JavaCore.newContainerEntry(
        JUnitCore.JUNIT4_CONTAINER_PATH,
        new IAccessRule[0],
        new IClasspathAttribute[] {nonDependencyAttribute},
        false);
    IClasspathEntry[] rawClasspath = javaProject.getRawClasspath();
    IClasspathEntry[] newRawClasspath = Arrays.copyOf(rawClasspath, rawClasspath.length + 1);
    newRawClasspath[newRawClasspath.length - 1] = junit4Container;
    javaProject.setRawClasspath(newRawClasspath, monitor);
  }
```

The line that fails is  javaProject.getRawClasspath() so if JavaCore.create(newProject) has not yet finished we could see this.

**Comment by elharo** [comment]
I have not been immediately able to reproduce this. Perhaps a thread issue?

```
  private static void addJunit4ToClasspath(IProgressMonitor monitor, IProject newProject)
      throws CoreException {
    IJavaProject javaProject = JavaCore.create(newProject);
    IClasspathAttribute nonDependencyAttribute =
        UpdateClasspathAttributeUtil.createNonDependencyAttribute();
    IClasspathEntry junit4Container = JavaCore.newContainerEntry(
        JUnitCore.JUNIT4_CONTAINER_PATH,
        new IAccessRule[0],
        new IClasspathAttribute[] {nonDependencyAttribute},
        false);
    IClasspathEntry[] rawClasspath = javaProject.getRawClasspath();
    IClasspathEntry[] newRawClasspath = Arrays.copyOf(rawClasspath, rawClasspath.length + 1);
    newRawClasspath[newRawClasspath.length - 1] = junit4Container;
    javaProject.setRawClasspath(newRawClasspath, monitor);
  }
```

The line that fails is  javaProject.getRawClasspath() so if JavaCore.create(newProject) has not yet finished we could see this.

**Comment by elharo** [comment]
This is the code that threw the exception. The problem is not that the project doesn't exist but that it doesn't have the Java nature:

```
	public PerProjectInfo getPerProjectInfoCheckExistence(IProject project) throws JavaModelException {
		JavaModelManager.PerProjectInfo info = getPerProjectInfo(project, false /* don't create info */);
		if (info == null) {
			if (!JavaProject.hasJavaNature(project)) {
				throw ((JavaProject)JavaCore.create(project)).newNotPresentException();
			}
			info = getPerProjectInfo(project, true /* create info */);
		}
		return info;
	}
```

**Comment by elharo** [comment]
It appears that in this case the project files weren't created before the exception was thrown. I don't immediately see what in this path could be asynchronous, but if it is then we could get to `addJunit4ToClasspath` before the project has been created and see these symptoms.  

```
    CreateProjectOperation operation = new CreateProjectOperation(description, operationLabel); 
    try {
      operation.execute(subMonitor.newChild(10), uiInfoAdapter);
      mostImportant = createProjectFiles(newProject, config, subMonitor.newChild(80));
    } catch (ExecutionException ex) {
      throw new InvocationTargetException(ex);
    }

    addAppEngineFacet(newProject, subMonitor.newChild(4));

    BuildPath.addLibraries(newProject, config.getAppEngineLibraries(), subMonitor.newChild(2));

    addJunit4ToClasspath(subMonitor.newChild(2), newProject);
```

**Comment by elharo** [comment]
FYI, facets were not added to the project either.

**Comment by chanseokoh** [comment]
I can reliably reproduce this when I comment out a few things between `operation.execute()` and `addJunit4ToClassPath()`:

```java
    CreateProjectOperation operation = new CreateProjectOperation(description, operationLabel);
    try {
      operation.execute(subMonitor.newChild(10), uiInfoAdapter);
      //mostImportant = createProjectFiles(newProject, config, subMonitor.newChild(80));
    } catch (ExecutionException ex) {
      throw new InvocationTargetException(ex);
    }

    //addAppEngineFacet(newProject, subMonitor.newChild(4));

    //BuildPath.addLibraries(newProject, config.getAppEngineLibraries(), subMonitor.newChild(2));

    addJunit4ToClasspath(subMonitor.newChild(2), newProject);
```

![event details _003](https://cloud.githubusercontent.com/assets/10523105/23799923/4fa881c0-0578-11e7-9427-0efa44f73d0b.png)

It may really be that `operation.execute()` does not run synchronously. Grabbing a workspace lock while doing `addJunit4ToClasspath()` might solve the problem. Not certain.

**Comment by elharo** [comment]
Interesting. I don't see any reason that addJunit4ToClasspath should fail without those commented out lines (unless addAppEngineFacet also adds the Java nature?). If that's true, we may be relying on the extra time those operations take to prevent a race condition.

I'd still like to understand why operation.execute is async. I don't see that yet.

**Comment by elharo** [comment]
OK, it looks like execute may be async:

	@Override
	public IStatus execute(IProgressMonitor monitor, final IAdaptable uiInfo)
			throws ExecutionException {
		try {
			getWorkspace().run(new IWorkspaceRunnable() {
				@Override
				public void run(IProgressMonitor monitor) throws CoreException {
					doExecute(monitor, uiInfo);
				}
			}, getExecuteSchedulingRule(), IWorkspace.AVOID_UPDATE, monitor);
		} catch (final CoreException e) {
			throw new ExecutionException(NLS.bind(
					UndoMessages.AbstractWorkspaceOperation_ExecuteErrorTitle,
					getLabel()), e);
		}
		isValid = true;
		return Status.OK_STATUS;
	}

In particular, getWorkspace.run():


Runs the given action as an atomic workspace operation.

After running a method that modifies resources in the workspace, registered listeners receive after-the-fact notification of what just transpired, in the form of a resource change event. This method allows clients to call a number of methods that modify resources and only have resource change event notifications reported at the end of the entire batch.

If this method is called outside the dynamic scope of another such call, this method runs the action and then reports a single resource change event describing the net effect of all changes done to resources by the action.

If this method is called in the dynamic scope of another such call, this method simply runs the action.

The supplied scheduling rule is used to determine whether this operation can be run simultaneously with workspace changes in other threads. If the scheduling rule conflicts with another workspace change that is currently running, the calling thread will be blocked until that change completes. If the action attempts to make changes to the workspace that were not specified in the scheduling rule, it will fail. If no scheduling rule is supplied, there are no scheduling restrictions for this operation. If a non-null scheduling rule is supplied, this operation must always support cancelation in the case where this operation becomes blocked by a long running background operation.

The AVOID_UPDATE flag controls whether periodic resource change notifications should occur during the scope of this call. If this flag is specified, and no other threads modify the workspace concurrently, then all resource change notifications will be deferred until the end of this call. If this flag is not specified, the platform may decide to broadcast periodic resource change notifications during the scope of this call.

Flags other than AVOID_UPDATE are ignored.

**Comment by chanseokoh** [comment]
Another possibility could be other things are synchronous, while the completion of adding a Java nature is async.

**Comment by briandealwis** [comment]
I'm pretty sure `IWorkspace.run()` is synchronous. The java nature should be added by the `addAppEngineFacets()` with the Java facet, which I believe should be added synchronously too.

**Comment by briandealwis** [comment]
I wonder if the actual hookup in JDT is done via the auto builder... That's something odd I've seen recently in #1529

**Comment by chanseokoh** [comment]
Explicitly setting the Java nature with `description.setNatureIds(new String[]{JavaCore.NATURE_ID});` when creating a project (i.e., setting it before `operation.execute()`) eliminates the error in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1568#issuecomment-285690915.

```java
    IProjectDescription description = workspace.newProjectDescription(name);
    description.setNatureIds(new String[]{JavaCore.NATURE_ID});  <------ HERE
    description.setLocationURI(location);

    String operationLabel = getDescription();
    SubMonitor subMonitor = SubMonitor.convert(monitor, operationLabel, 100);
    CreateProjectOperation operation = new CreateProjectOperation(description, operationLabel);
    try {
      operation.execute(subMonitor.newChild(10), uiInfoAdapter);
      //mostImportant = createProjectFiles(newProject, config, subMonitor.newChild(80));
    } catch (ExecutionException ex) {
      throw new InvocationTargetException(ex);
    }

    //addAppEngineFacet(newProject, subMonitor.newChild(4));

    //BuildPath.addLibraries(newProject, config.getAppEngineLibraries(), subMonitor.newChild(2));

    addJunit4ToClasspath(subMonitor.newChild(2), newProject);
```

It might be the case that either `addAppEngineFacet()` or `BuildPath.addLibraries()` somehow adds the Java nature (or something similar to that effect) but it does it in an async way.

**Comment by chanseokoh** [comment]
> The java nature should be added by the addAppEngineFacets() with the Java facet, which I believe should be added synchronously too.

So, I think it's not robust to rely on that `addAppEngineFacets()` will add the Java nature. If we don't call it, `addJunit4ToClasspath()` will fail. (Not to mention we are not 100% certain if it will add the nature synchronously). So I think it's best to explicitly sett the nature when creating a project initially.

**Comment by chanseokoh** [comment]
> FYI, facets were not added to the project either.

This concerns me a bit. @elharo If you still have app that failed, can you check the content of the facet metadata conf file? If you did in a hosted Eclipse dev run, it should be at (if the app was named `test1`)

`$HOME/runtime-gcloud-eclipse/test1/.settings/org.eclipse.wst.common.project.facet.core.xml`

**Comment by elharo** [comment]
Good question. The only thing that was created was the .project file. Nothing else That contains

```
<?xml version="1.0" encoding="UTF-8"?>
<projectDescription>
        <name>test1</name>
        <comment></comment>
        <projects>
        </projects>
        <buildSpec>
        </buildSpec>
        <natures>
        </natures>
</projectDescription>
```

**Comment by chanseokoh** [comment]
> The java nature should be added by the addAppEngineFacets() with the Java facet, which I believe should be added synchronously too.

@briandealwis was right. Adding the Java 1.7 facet adds the Java nature (within its delegate `JavaFacetInstallDelegate.execute()`),

```java
            final IProjectDescription desc = project.getDescription();
            final String[] current = desc.getNatureIds();
            final String[] replacement = new String[ current.length + 1 ];
            System.arraycopy( current, 0, replacement, 0, current.length );
            replacement[ current.length ] = JavaCore.NATURE_ID;
            desc.setNatureIds( replacement );
            project.setDescription( desc, null );
```

and it does it synchronously:

![selection_004](https://cloud.githubusercontent.com/assets/10523105/23812590/23b6bce0-05a9-11e7-9266-6e073d7c3ec1.png)

The only reason for the error I can think of now is that `addAppEngineFacet()` failed or canceled somehow mysteriously. The fact that there was no facet metatadata file also supports that.

**Comment by elharo** [comment]
It didn't create the project files either.

**Comment by briandealwis** [comment]
Can you attach the .../.metadata/.log? Was this is a new workspace? Is it possible the workspace already had a `test1`?

**Comment by elharo** [comment]
Reasonably confident there was no pre-existing test1 project in the workspace.  This is what's in the .log ATM though I suspect it's been overwritten several times since:

!SESSION 2016-12-02 00:38:01.004 -----------------------------------------------
eclipse.buildId=unknown
java.version=1.7.0_79
java.vendor=Oracle Corporation
BootLoader constants: OS=macosx, ARCH=x86_64, WS=cocoa, NL=en_US
Framework arguments:  -product org.eclipse.platform.ide
Command-line arguments:  -product org.eclipse.platform.ide -data /Users/elharo/google-cloud-eclipse-workspace/../runtime-gcloud-eclipse -dev file:/Users/elharo/google-cloud-eclipse-workspace/.metadata/.plugins/org.eclipse.pde.core/gcloud-eclipse-tools/dev.properties -os macosx -ws cocoa -arch x86_64 -consoleLog

!ENTRY org.eclipse.core.net 1 0 2016-12-02 00:38:16.044
!MESSAGE System property http.nonProxyHosts has been set to local|*.local|169.254/16|*.169.254/16 by an external source. This value will be overwritten using the values from the preferences

!ENTRY org.eclipse.wst.server.discovery 1 0 2016-12-02 00:38:25.404
!MESSAGE cacheFile=file:/Users/elharo/google-cloud-eclipse-workspace/.metadata/.plugins/org.eclipse.pde.core/gcloud-eclipse-tools/org.eclipse.wst.server.discovery/serverAdapterSitesDetails.xml

!ENTRY org.eclipse.core.jobs 2 2 2016-12-02 00:38:45.353
!MESSAGE Job found still running after platform shutdown.  Jobs should be canceled by the plugin that scheduled them during shutdown: org.eclipse.wst.server.ui.ServerUIUtil$1RefreshJob

**Comment by briandealwis** [comment]
Hmmm that doesn't look right: check out the dates.

**Comment by elharo** [comment]
You're right. Wrong workspace. I think this is the right one.
[dot log.txt](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/files/836466/dot.log.txt)

**Comment by elharo** [comment]
Looks like Maven may have had problems:


!ENTRY org.eclipse.osgi 2 0 2017-03-10 07:29:28.427
!MESSAGE While loading class "org.eclipse.m2e.core.project.IMavenProjectRegistry", thread "Thread[Worker-7,5,main]" timed out waiting (5076ms) for thread "Thread[org.eclipse.wst.sse.ui.internal.reconcile.StructuredRegionProcessor,1,main]" to finish starting bundle "org.eclipse.m2e.core_1.6.2.20150902-0002 [484]". To avoid deadlock, thread "Thread[Worker-7,5,main]" is proceeding but "org.eclipse.m2e.core.project.IMavenProjectRegistry" may not be fully initialized.


!ENTRY org.eclipse.osgi 2 0 2017-03-10 07:29:33.464
!MESSAGE While loading class "org.eclipse.m2e.core.project.IMavenProjectChangedListener", thread "Thread[Worker-7,5,main]" timed out waiting (5003ms) for thread "Thread[org.eclipse.wst.sse.ui.internal.reconcile.StructuredRegionProcessor,1,main]" to finish starting bundle "org.eclipse.m2e.core_1.6.2.20150902-0002 [484]". To avoid deadlock, thread "Thread[Worker-7,5,main]" is proceeding but "org.eclipse.m2e.core.project.IMavenProjectChangedListener" may not be fully initialized.
!STACK 0
org.osgi.framework.BundleException: Unable to acquire the state change lock for the module: osgi.identity; osgi.identity="org.eclipse.m2e.core"; type="osgi.bundle"; version:Version="1.6.2.20150902-0002"; singleton:="true" [id=484] STARTED [STARTED]
	at org.eclipse.osgi.container.Module.lockStateChange(Module.java:329)
	at org.eclipse.osgi.container.Module.start(Module.java:389)
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:454)
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.pos

**Comment by elharo** [comment]
Does a late failure of CreateAppEngineProject.execute cause a rollback or failure-to-commit of the entire workspace modify operation? If so, this might explain what I saw. I.e. the files were created but since BuildPath.addLibraries threw an exception, the files were undone?

**Comment by chanseokoh** [comment]
We create all project files in the separate `createProjectFiles()` after `execute()` succeeds, so I think it is not likely that that `execute()` will roll back that.

Anyway, the error log sounds like a serious internal framework error to me. Looks like it failed to load an M2E bundle, which shouldn't happen.

**Comment by elharo** [comment]
This hasn't recurred so my current thinking is that, unless someone sees something obvious and a way to fix it, we should move forward with the next release but leave this bug open for the time being.

**Comment by elharo** [comment]
May have just reproduced this in a flex project:

Failed to create project: flexme does not exist

Or it could be a bug in the flex new project wizard. Possibly this happens the first time you create a project in a new workspace (which would explain why we couldn't reproduce it before). 

![image](https://cloud.githubusercontent.com/assets/1005544/23849800/4f2c4212-07b3-11e7-9a18-4ac546dd1811.png)

However "An exception stack trace is not available."

**Comment by chanseokoh** [comment]
It's no wonder the flex wizard fails, because it never adds the Java nature. (In the flex case, `addAppEngineFacet()` is empty and does nothing, whereas in the standard case, the method adds the Java nature as I said in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1568#issuecomment-285781608.)

**Comment by briandealwis** [comment]
The bundle-timeout errors seems to be a phantom error that nobody has been able to reproduce (bugs [387080](https://bugs.eclipse.org/bugs/show_bug.cgi?id=387080), [387081](https://bugs.eclipse.org/bugs/show_bug.cgi?id=387081), and [424671](https://bugs.eclipse.org/bugs/show_bug.cgi?id=424671)).  From the looks of it, the timeout came from loading one of the classpath container initializers.  It might be worth enabling the _org.eclipse.jdt.core/debug/cpresolution/advanced_ tracing element (in the launch configuration > Configuration >  — that should dump out what initializer is being resolved (viz `JavaCore#computeClasspathContainerInitializer()`).  If this is the cause, a possible workaround might be for our wizards to explicit start the `org.eclipse.m2e.core` bundle.  

But that said, the actual error time new-project error time is 07:32:05.685, which is 2.5 minutes after these bundle-timeouts.  It's possible the bundles were left in an inconsistent state.

**Comment by diegomarquezp** [comment]
Closing as not planned since GAE with Java 8 is not supported. cc: @blakeli0
